File: debug_rsa.py
"""Debug: test both custom RSA and pycryptodome RSA against known test vector"""
import requests, urllib3, re, random, base64, struct
urllib3.disable_warnings()

USERNAME = "请替换"
PASSWORD = "请替换"

# ====== Get real modulus from server ======
s = requests.Session()
s.headers.update({'User-Agent': 'Mozilla/5.0 Chrome/150.0.0.0'})
s.verify = False
r = s.get('https://xg2.nsmc.edu.cn/Sys/UserLogin.aspx', timeout=15)
m = re.search(r'new RSAKeyPair\("([^"]+)",\s*"([^"]*)",\s*"([^"]+)"\)', r.text)
modulus = m.group(3)
print(f'Modulus len={len(modulus)}')

# ====== Base plaintext ======
raw = base64.b64encode(USERNAME.encode()).decode() + '\\' + base64.b64encode(PASSWORD.encode()).decode()
raw_bytes = raw.encode()
print(f'Plaintext ({len(raw_bytes)} bytes): {raw}')

n = int(modulus, 16)
e = 0x010001
digit_size = 128
msg_len = len(raw_bytes)

# ====== Method A: custom padding (little-endian) ======
padded = max(8, digit_size - 3 - msg_len)
b = bytearray(digit_size)
for x in range(msg_len): b[x] = raw_bytes[msg_len - 1 - x]
b[msg_len] = 0
for x in range(padded): b[msg_len + 1 + x] = random.randint(1, 255)
b[digit_size - 2] = 2
b[digit_size - 1] = 0
custom_int = int.from_bytes(b, 'little')
custom_enc = hex(pow(custom_int, e, n))[2:].zfill(256)
print(f'\nMethod A (custom little-endian): hex_len={len(custom_enc)}')

# ====== Method B: pycryptodome PKCS1-v1.5 ======
from Crypto.PublicKey import RSA as CRSA
from Crypto.Cipher import PKCS1_v1_5 as CPKCS

# Build RSA public key
rsa_key = CRSA.construct((n, e))
cipher = CPKCS.new(rsa_key)
# In PKCS1-v1.5 type 2, cipher.encrypt() handles padding automatically
ciphertext = cipher.encrypt(raw_bytes)
b64_enc = base64.b64encode(ciphertext).decode()
pycrypto_hex = ciphertext.hex()
print(f'Method B (pycryptodome PKCS1-v1.5): hex_len={len(pycrypto_hex)}')
print(f'  b64: {b64_enc}')

# ====== Try both on server ======
s2 = requests.Session()
s2.headers.update({'User-Agent': 'Mozilla/5.0 Chrome/150.0.0.0'})
s2.verify = False
s2.get('https://xg2.nsmc.edu.cn/Sys/UserLogin.aspx', timeout=10)

cap = ''.join(random.choice(['2','3','4','5','6','7','8','9','b','c','e','f','g','h','j','k','m','n','p','r','s','t','u','v','w','x','y','z','B','C','E','F','G','H','J','K','M','N','P','R','S','T','U','V','W','X','Y','Z']) for _ in range(4))

def try_login(session, encrypted_hex, label):
    """Try login with a given encrypted value"""
    # Get fresh fields from the page we already have
    cap2 = ''.join(random.choice(['2','3','4','5','6','7','8','9','b','c','e','f','g','h','j','k','m','n','p','r','s','t','u','v','w','x','y','z','B','C','E','F','G','H','J','K','M','N','P','R','S','T','U','V','W','X','Y','Z']) for _ in range(4))

    fd = {
        '__LASTFOCUS': '', '__EVENTTARGET': '', '__EVENTARGUMENT': '',
        '__VIEWSTATE': re.search(r'__VIEWSTATE[^>]*value="([^"]*)"', r.text).group(1) if re.search(r'__VIEWSTATE[^>]*value="([^"]*)"', r.text) else '',
        '__VIEWSTATEGENERATOR': re.search(r'__VIEWSTATEGENERATOR[^>]*value="([^"]*)"', r.text).group(1) if re.search(r'__VIEWSTATEGENERATOR[^>]*value="([^"]*)"', r.text) else '',
        '__VIEWSTATEENCRYPTED': '',
        '__EVENTVALIDATION': re.search(r'__EVENTVALIDATION[^>]*value="([^"]*)"', r.text).group(1) if re.search(r'__EVENTVALIDATION[^>]*value="([^"]*)"', r.text) else '',
        'UserName': '******', 'posx': encrypted_hex, 'codeInput': cap2,
        'queryBtn': '登          录',
    }

    import urllib.parse
    body = '&'.join(f"{urllib.parse.quote_from_bytes(k.encode('gb2312'), safe='')}={urllib.parse.quote_from_bytes(v.encode('gb2312'), safe='')}" for k,v in fd.items())

    resp = session.post('https://xg2.nsmc.edu.cn/Sys/UserLogin.aspx', data=body,
                        headers={'Content-Type': 'application/x-www-form-urlencoded'}, timeout=15)
    h = resp.text
    if 'Navigation.aspx' in h:
        print(f'  [{label}] ✅ SUCCESS!')
        return True
    if 'MainFrame.aspx' in h:
        print(f'  [{label}] ✅ SUCCESS (MainFrame)!')
        return True
    if 'StuLeave' in h:
        print(f'  [{label}] ✅ SUCCESS (Leave page mention)!')
        return True
    alert = re.search(r"alert\(['\"]([^'\"]+)['\"]\)", h)
    if alert:
        print(f'  [{label}] Server alert: {alert.group(1)}')
    elif 'IndexOutOfRangeException' in h or '索引超出了数组' in h:
        print(f'  [{label}] ❌ IndexOutOfRangeException')
    else:
        print(f'  [{label}] ❌ Unknown, len={len(h)} vs login page 10758')
        print(f'  Contains Navigation.aspx: {"Navigation.aspx" in h}')
        print(f'  Contains MainFrame.aspx: {"MainFrame.aspx" in h}')
        print(f'  Contains queryBtn: {"queryBtn" in h}')
        # look for layer.alert
        layer = re.search(r'layer\.alert\([\'"]([^\'"]+)[\'"]\)', h)
        if layer:
            print(f'  layer.alert: {layer.group(1)}')
        with open(f'xg2_resp_{label}.html', 'w', encoding='utf-8') as f:
            f.write(h)
    return False

# Try both methods
try_login(s2, custom_enc, 'Custom')
try_login(s2, pycrypto_hex, 'Pycryptodome')

File: debug_rsa.js
// 用 JS RSA 加密来验证 Python 实现是否正确
const https = require('https');
const forge = require('node-forge');
const b64 = require('base-64');

const username = process.argv[2] || 'test';
const password = process.argv[3] || 'test';

function getLoginPage() {
    return new Promise((resolve, reject) => {
        https.get('https://xg2.nsmc.edu.cn/Sys/UserLogin.aspx', {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/150.0.0.0 Safari/537.36 Edg/150.0.0.0'
            },
            rejectUnauthorized: false
        }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => resolve(data));
        }).on('error', reject);
    });
}

async function main() {
    console.log('Username:', username);
    console.log('Password:', password);

    const html = await getLoginPage();
    const m = html.match(/new RSAKeyPair\("([^"]+)",\s*"([^"]*)",\s*"([^"]+)"\)/);
    if (!m) { console.log('RSA not found'); return; }
    const [_, exp, zero, mod] = m;
    console.log('Modulus:', mod.substring(0, 40) + '...');
    console.log('Modulus length:', mod.length);

    // ====== 方法1: 使用 node-forge 标准 PKCS1 v1.5 ======
    const n = new forge.jsbn.BigInteger(mod, 16);
    const e = new forge.jsbn.BigInteger('010001', 16);
    const rsaKey = forge.pki.setRsaPublicKey(n, e);

    const raw = b64.encode(username) + '\\' + b64.encode(password);
    console.log('Raw:', raw);

    // 标准 PKCS1-v1.5 加密
    const encrypted = rsaKey.encrypt(raw, 'RSAES-PKCS1-V1_5');
    const hex_std = forge.util.bytesToHex(encrypted);
    console.log('forge hex:', hex_std.substring(0, 60) + '...');
    console.log('forge hex len:', hex_std.length);
    console.log('forge b64:', forge.util.encode64(encrypted));

    // ====== 方法2: 直接 pow_mod (与 Python 一致) ======
    const rawBytes = Buffer.from(raw, 'ascii');
    console.log('\nraw bytes hex:', rawBytes.toString('hex'));
    console.log('raw bytes length:', rawBytes.length);

    // 构建与 Python 相同的消息块
    const digit_size = 128;
    const msg_len = rawBytes.length;
    const padded_size = Math.max(8, digit_size - 3 - msg_len);

    const b = Buffer.alloc(digit_size, 0);
    for (let x = 0; x < msg_len; x++) {
        b[x] = rawBytes[msg_len - 1 - x];
    }
    b[msg_len] = 0;
    for (let x = 0; x < padded_size; x++) {
        b[msg_len + 1 + x] = Math.floor(Math.random() * 254) + 1;
    }
    b[digit_size - 2] = 2;
    b[digit_size - 1] = 0;

    console.log('padded hex:', b.toString('hex'));
    console.log('padded length:', b.length);
    console.log('padded first 4 bytes:', b.slice(0, 4).toString('hex'));
    console.log('padded last 4 bytes:', b.slice(-4).toString('hex'));

    // 转为 big-endian 整数 (Python 用 little-endian, JS 用 big-endian)
    const msgInt = new forge.jsbn.BigInteger(b.toString('hex'), 16);
    const encrypted2 = msgInt.modPow(e, n);
    const hex_pow = encrypted2.toString(16);
    console.log('\npow_mod hex:', hex_pow.substring(0, 60) + '...');
    console.log('pow_mod hex len:', hex_pow.length);

    // 现在用我们 Python 的 little-endian 方式计算
    const leInt = BigInt('0x' + b.toString('hex'));
    console.log('\nPure Python expected:');
    const n_big = BigInt('0x' + mod);
    const e_big = 65537n;
    const result = leInt ** e_big % n_big;
    const resultHex = result.toString(16);
    console.log('result hex:', resultHex.substring(0, 40) + '...');
    console.log('result hex len:', resultHex.length);
    console.log('zfilled:', resultHex.padStart(256, '0').substring(0, 40) + '...');
}

main().catch(console.error);
